File: recomp-mobile/app/src/main/java/com/refactor/app/ui/workouts/WorkoutImportStart.kt
package com.refactor.app.ui.workouts

import com.refactor.app.api.dto.WorkoutDayDto
import java.time.DayOfWeek
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.time.temporal.TemporalAdjusters
import java.util.Locale
import java.util.regex.Pattern

/** Matches web `workout-import-start.ts`. */
object WorkoutImportStart {
    private val isoDate = DateTimeFormatter.ISO_LOCAL_DATE
    private val weekdayNames = listOf("sunday", "monday", "tuesday", "wednesday", "thursday", "friday", "saturday")
    private val shortNames = listOf("sun", "mon", "tue", "wed", "thu", "fri", "sat")
    private val weekPattern = Pattern.compile("week\\s*(\\d+)", Pattern.CASE_INSENSITIVE)

    /** 0 = Sunday … 6 = Saturday (matches JavaScript `Date.getDay()`). */
    fun weekdayIndex(dayLabel: String): Int? {
        val lower = dayLabel.trim().lowercase(Locale.US)
        weekdayNames.forEachIndexed { i, name ->
            if (lower == name || lower.startsWith("$name ") || lower.startsWith(shortNames[i])) return i
        }
        return null
    }

    fun extractProgramWeek(dayLabel: String): Int? {
        val m = weekPattern.matcher(dayLabel)
        return if (m.find()) m.group(1)?.toIntOrNull() else null
    }

    fun isAnchoredProgram(weeklyPlan: List<WorkoutDayDto>): Boolean =
        weeklyPlan.size > 7 || weeklyPlan.any { extractProgramWeek(it.day) != null }

    fun nextOccurrenceOfWeekday(weekdayIndex: Int, today: LocalDate = LocalDate.now()): LocalDate {
        val todayDow = today.dayOfWeek.value % 7 // Sun=0 … Sat=6
        val daysUntil = (weekdayIndex - todayDow + 7) % 7
        return today.plusDays(daysUntil.toLong())
    }

    fun inferFirstSessionDate(weeklyPlan: List<WorkoutDayDto>, today: LocalDate = LocalDate.now()): LocalDate {
        val first = weeklyPlan.firstOrNull { it.day.isNotBlank() } ?: return nextOccurrenceOfWeekday(1, today)
        val wd = weekdayIndex(first.day) ?: return nextOccurrenceOfWeekday(1, today)
        return nextOccurrenceOfWeekday(wd, today)
    }

    fun inferProgramWeek1Start(weeklyPlan: List<WorkoutDayDto>, today: LocalDate = LocalDate.now()): String? {
        if (weeklyPlan.isEmpty() || !isAnchoredProgram(weeklyPlan)) return null
        val firstSession = inferFirstSessionDate(weeklyPlan, today)
        val monday = firstSession.with(TemporalAdjusters.previousOrSame(DayOfWeek.MONDAY))
        return monday.format(isoDate)
    }
}

File: RecompSwift/RefactorKit/Sources/Utilities/WorkoutImportStart.swift
import Foundation

/// Matches web `workout-import-start.ts` — anchor imported / multi-week programs to the next session weekday.
public enum WorkoutImportStart {
    private static let weekdayNames = ["sunday", "monday", "tuesday", "wednesday", "thursday", "friday", "saturday"]
    private static let shortNames = ["sun", "mon", "tue", "wed", "thu", "fri", "sat"]

    /// 0 = Sunday … 6 = Saturday (matches JavaScript `Date.getDay()`).
    public static func weekdayIndex(from dayLabel: String) -> Int? {
        let lower = dayLabel.lowercased().trimmingCharacters(in: .whitespacesAndNewlines)
        for (i, name) in weekdayNames.enumerated() {
            if lower == name || lower.hasPrefix("\(name) ") || lower.hasPrefix(shortNames[i]) {
                return i
            }
        }
        return nil
    }

    public static func extractProgramWeek(from dayLabel: String) -> Int? {
        let pattern = #"week\s*(\d+)"#
        guard let regex = try? NSRegularExpression(pattern: pattern, options: .caseInsensitive),
              let match = regex.firstMatch(in: dayLabel, range: NSRange(dayLabel.startIndex..., in: dayLabel)),
              let range = Range(match.range(at: 1), in: dayLabel),
              let n = Int(dayLabel[range]) else { return nil }
        return n
    }

    public static func isAnchoredProgram(_ weeklyPlan: [WorkoutDay]) -> Bool {
        weeklyPlan.count > 7 || weeklyPlan.contains { extractProgramWeek(from: $0.day) != nil }
    }

    public static func nextOccurrence(ofWeekday weekdayIndex: Int, today: String = DateHelpers.todayString()) -> String {
        guard let todayDate = DateHelpers.date(from: today) else { return today }
        let cal = Calendar.current
        let noon = cal.date(bySettingHour: 12, minute: 0, second: 0, of: todayDate) ?? todayDate
        let todayDow = cal.component(.weekday, from: noon) - 1
        let daysUntil = (weekdayIndex - todayDow + 7) % 7
        guard let next = cal.date(byAdding: .day, value: daysUntil, to: noon) else { return today }
        return DateHelpers.dateString(from: next)
    }

    public static func inferFirstSessionDate(weeklyPlan: [WorkoutDay], today: String = DateHelpers.todayString()) -> String {
        let first = weeklyPlan.first(where: { !$0.day.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }) ?? weeklyPlan.first
        guard let day = first else { return nextOccurrence(ofWeekday: 1, today: today) }
        if let wd = weekdayIndex(from: day.day) {
            return nextOccurrence(ofWeekday: wd, today: today)
        }
        return nextOccurrence(ofWeekday: 1, today: today)
    }

    public static func inferProgramWeek1Start(weeklyPlan: [WorkoutDay], today: String = DateHelpers.todayString()) -> String? {
        guard !weeklyPlan.isEmpty, isAnchoredProgram(weeklyPlan) else { return nil }
        return DateHelpers.mondayWeekStartString(
            containingCalendarDay: inferFirstSessionDate(weeklyPlan: weeklyPlan, today: today)
        )
    }

    public static func workoutPlanAfterImport(weeklyPlan: [WorkoutDay], preserving existing: WorkoutPlan) -> WorkoutPlan {
        var wp = existing
        wp.weeklyPlan = weeklyPlan
        if let anchor = inferProgramWeek1Start(weeklyPlan: weeklyPlan) {
            wp.programWeek1Start = anchor
            wp.programWeekOffset = 0
            wp.missedSessions = []
            wp.catchUpBannerDismissedAt = nil
        } else if weeklyPlan.count <= 7 {
            wp.programWeek1Start = nil
        }
        return wp
    }
}
